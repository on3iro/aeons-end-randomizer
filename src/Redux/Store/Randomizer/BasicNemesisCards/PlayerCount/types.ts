import { ActionsUnion } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { actions } from './actions'

export type State = PlayerCount

export enum ActionTypes {
  SET = 'BasicNemesisCards/PlayerCount/SET',
}

export type Action = ActionsUnion<typeof actions>

export type PlayerCount = types.PlayerCount

export type PlayerCountStateSlice = {
  Randomizer: {
    BasicNemesisCards: {
      PlayerCount: PlayerCount
    }
  }
}
