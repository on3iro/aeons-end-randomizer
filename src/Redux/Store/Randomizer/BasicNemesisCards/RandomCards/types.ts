import { ActionsUnion } from '@martin_hotell/rex-tils'

import { actions } from './actions'

export type State = Readonly<{
  Tier1: ReadonlyArray<Readonly<{ id: string }>>
  Tier2: ReadonlyArray<Readonly<{ id: string }>>
  Tier3: ReadonlyArray<Readonly<{ id: string }>>
  recoverySeed?: string
}>

export enum ActionTypes {
  CREATE = 'BasicNemesisCards/RandomCards/CREATE',
}

export type Action = ActionsUnion<typeof actions>

export type RandomCardsStateSlice = {
  Randomizer: {
    BasicNemesisCards: {
      RandomCards: State
    }
  }
}
