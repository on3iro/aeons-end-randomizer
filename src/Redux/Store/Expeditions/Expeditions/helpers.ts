import shortid from 'shortid'

import * as types from 'types'
import { byCost } from 'helpers'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createSupply,
} from 'Redux/helpers'

import { BaseConfig, RollBattleConfig, WinConfig, LossConfig } from './types'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}

export const determineBaseTier = () => {}

export const generateBattles = (
  variant: types.Variant,
  expeditionId: string
) => {
  const battles = variant.configList.map(
    (config, index): types.Battle => {
      const isFirst = index === 0

      return {
        id: shortid.generate(),
        nemesisTier: config.tier,
        treasure: config.treasure,
        expeditionId,
        status: isFirst ? 'unlocked' : 'locked',
        tries: 0,
      }
    }
  )

  return battles
}

export const createExpeditionConfig = ({
  variant,
  name,
  bigPocketVariant,
  availableMageIds,
  availableCards,
  availableLevel1TreasureIds,
  tiles,
}: BaseConfig) => {
  const mageIds = createIdList(
    availableMageIds,
    createArrayWithDefaultValues(4, 'EMPTY'),
    getRandomEntity
  ).result
  const { gems, relics, spells } = createSupply(availableCards, tiles)
  const gemsByCost = gems.result.sort(byCost)
  const relicsByCost = relics.result.sort(byCost)
  const spellsByCost = spells.result.sort(byCost)

  const supplyIds = [...gemsByCost, ...relicsByCost, ...spellsByCost].map(
    card => card.id
  )

  const startsWithTreasure = variant.configList[0].tier.tier > 1 // TODO should probably use our rollTreasure property
  const treasureIds = startsWithTreasure
    ? createIdList(
        availableLevel1TreasureIds,
        createArrayWithDefaultValues(5, 'EMPTY'),
        getRandomEntity
      ).result
    : []

  const expeditionId = shortid.generate()
  const battles = generateBattles(variant, expeditionId)

  return {
    battles,
    bigPocketVariant,
    expeditionId,
    mageIds,
    name,
    supplyIds,
    treasureIds,
    variantId: variant.id,
  }
}

// Because we always add these cards to an existing array inside our expedition in our store,
// the count of newly added cards decreases by tier.
// If we ever have a variant, that starts higher than tier 2 we have to
// incorporate changes to roll the whole amount of cards per tier!
const getUpgradedBasicNemesisIdsByBattleTier = ({
  availableTier1Ids,
  availableTier2Ids,
  availableTier3Ids,
  battleTier,
}: {
  availableTier1Ids: string[]
  availableTier2Ids: string[]
  availableTier3Ids: string[]
  battleTier: 1 | 2 | 3 | 4
}) => {
  switch (battleTier) {
    case 1: {
      return [] // No upgraded cards are added on tier 1
    }

    case 2: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    case 3:
    case 4: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(2, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    default: {
      return [] // Should never occur!
    }
  }
}

const rollNewUpgradedNemesisCards = (
  availableUpgradedBasicNemesisCards: types.UpgradedBasicNemesisCard[],
  previousUpgradedBasicNemesisCards: string[],
  nemesisTier: 1 | 2 | 3 | 4
) => {
  const upgradedCardsWithoutPreviousCards = availableUpgradedBasicNemesisCards.filter(
    upgradedCard => !previousUpgradedBasicNemesisCards.includes(upgradedCard.id)
  )

  const tier1AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 1)
    .map(card => card.id)
  const tier2AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 2)
    .map(card => card.id)
  const tier3AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 3)
    .map(card => card.id)

  const upgradedBasicNemesisCardIds = getUpgradedBasicNemesisIdsByBattleTier({
    battleTier: nemesisTier,
    availableTier1Ids: tier1AvailableUpgradedNemesisIds,
    availableTier2Ids: tier2AvailableUpgradedNemesisIds,
    availableTier3Ids: tier3AvailableUpgradedNemesisIds,
  })

  return [...previousUpgradedBasicNemesisCards, ...upgradedBasicNemesisCardIds]
}

export const createBattle = (config: RollBattleConfig) => {
  const nemesisIds = config.availableNemeses
    .filter(
      nemesis => nemesis.expeditionRating === config.battle.nemesisTier.tier
    )
    .map(nemesis => nemesis.id)
    .filter(nemesisId => !config.previousNemeses.includes(nemesisId))

  const nemesisId = createIdList(
    nemesisIds,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getRandomEntity
  ).result[0]

  const upgradedBasicNemesisCardIds = config.battle.nemesisTier.isNewTier
    ? rollNewUpgradedNemesisCards(
        config.availableUpgradedBasicNemesisCards,
        config.previousUpgradedBasicNemesisCards,
        config.battle.nemesisTier.tier
      )
    : config.previousUpgradedBasicNemesisCards

  return {
    battle: { ...config.battle, nemesisId, status: 'before_battle' },
    upgradedBasicNemesisCardIds,
  }
}

const rollNewEntity = (list: string[]): string =>
  createIdList(list, createArrayWithDefaultValues(1, 'EMPTY'), getRandomEntity)
    .result[0]

export const rollWinRewards = (config: WinConfig) => {
  const newTreasures = createIdList(
    config.treasureIds,
    createArrayWithDefaultValues(5, 'EMPTY'),
    getRandomEntity
  ).result

  const newGem = rollNewEntity(config.gemIds)
  const newRelic = rollNewEntity(config.relicIds)
  const newSpell = rollNewEntity(config.spellIds)

  const updatedBattle = {
    ...config.battle,
    rewards: {
      treasure: newTreasures,
      supplyIds: [newGem, newRelic, newSpell],
      mage: undefined, // we explicitely overwrite the mage reward from earlier losses
    },
  }
  return { battle: updatedBattle }
}

export const rollLossRewards = (config: LossConfig) => {
  switch (config.type) {
    case 'mage': {
      const newMage = rollNewEntity(config.mageIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: newMage,
          supplyIds: [],
        },
      }
    }

    case 'gem': {
      const newGem = rollNewEntity(config.gemIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newGem],
        },
      }
    }

    case 'relic': {
      const newRelic = rollNewEntity(config.relicIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newRelic],
        },
      }
    }

    case 'spell': {
      const newSpell = rollNewEntity(config.spellIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newSpell],
        },
      }
    }

    case 'treasure1': {
      const newTreasure = rollNewEntity(config.treasure1Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure2': {
      const newTreasure = rollNewEntity(config.treasure2Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure3': {
      const newTreasure = rollNewEntity(config.treasure3Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    default: {
      return { ...config.battle }
    }
  }
}
