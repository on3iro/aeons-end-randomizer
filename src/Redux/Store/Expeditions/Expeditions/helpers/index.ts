export * from './convertExpeditionToConfig'
export * from './convertExpeditionFromConfig'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}
