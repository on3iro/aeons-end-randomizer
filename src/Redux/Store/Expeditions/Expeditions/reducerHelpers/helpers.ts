import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'
import * as types from 'aer-types/types'

import { State } from '../types'
import { actions } from '../actions'

export const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

export const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: types.Rewards
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const branches = oldExpedition.sequence.branches

  const updatedBranches = {
    ...branches,
    [battle.id]: {
      ...battle,
      ...additionalBattleProps,
    },
  }

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        sequence: {
          ...oldExpedition.sequence,
          branches: updatedBranches,
        },
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
