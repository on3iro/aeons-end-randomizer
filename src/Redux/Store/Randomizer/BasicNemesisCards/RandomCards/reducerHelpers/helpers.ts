import seedrandom from 'seedrandom'
import shortid from 'shortid'

import * as types from 'types'

import { PlayerCount } from '../../PlayerCount/types'

import {
  getRandomEntity,
  createSlotList,
  createBasicNemesisCardList,
} from 'Redux/helpers'

type CardMatrix = {
  [key: string]: {
    [key: number]: number
  }
}

export const cardMatrix: CardMatrix = {
  tier1: {
    1: 1,
    2: 3,
    3: 5,
    4: 8,
  },
  tier2: {
    1: 3,
    2: 5,
    3: 6,
    4: 7,
  },
  tier3: {
    1: 7,
    2: 7,
    3: 7,
    4: 7,
  },
}

export const getBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier
) => cards.filter(card => card.tier === tier)

export const getRandomMinionAmount = (
  count: number = 3,
  seed: types.Seed = {
    seed: shortid.generate(),
  }
) => {
  const rng = seedrandom(seed.seed)
  return Math.floor(rng() * Math.floor(count))
}

export const drawMinions = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const minions = cards.filter(card => card.type === 'Minion')

  return createBasicNemesisCardList(
    minions,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result
}

export const drawPowersAndAttacks = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const powersAndAttacks = cards.filter(card => card.type !== 'Minion')

  return createBasicNemesisCardList(
    powersAndAttacks,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result
}

export const getRandomBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier,
  playerCount: PlayerCount,
  seed?: types.Seed
) => {
  // Get cards by tier
  const cardsByTier = getBasicNemesisCardsByTier(cards, tier)

  // Get amount of minions per player
  const minionAmount =
    playerCount === 1
      ? getRandomMinionAmount(2, seed)
      : getRandomMinionAmount(3, seed)

  // Draw up to max amount of minions from minions
  const randomMinions = drawMinions(cardsByTier, minionAmount, seed)

  // Draw tierCount - minionCount cards from power/attack pool
  const randomPowersAndAttacks = drawPowersAndAttacks(
    cardsByTier,
    cardMatrix[`tier${tier}`][playerCount] - randomMinions.length,
    seed
  )

  return [...randomMinions, ...randomPowersAndAttacks]
}

export const getRandomBasicNemesisCardList = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  playerCount: PlayerCount,
  seed?: types.Seed
) => {
  return {
    Tier1: getRandomBasicNemesisCardsByTier(cards, 1, playerCount, seed),
    Tier2: getRandomBasicNemesisCardsByTier(cards, 2, playerCount, seed),
    Tier3: getRandomBasicNemesisCardsByTier(cards, 3, playerCount, seed),
  }
}
