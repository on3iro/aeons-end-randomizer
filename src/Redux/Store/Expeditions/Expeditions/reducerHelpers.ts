import { loop, Cmd } from 'redux-loop'
import { get as getFromDb, set as setToDb } from 'idb-keyval'

import * as types from 'types'

import * as helpers from './helpers'
import * as sideEffects from './sideEffects'
import { State, Action } from './types'
import { actions } from './actions'

const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: null
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )

  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: {
      ...battle,
      ...additionalBattleProps,
    },
  })

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        battles: updatedBattles,
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const createExpedition = (
  state: State,
  action: ReturnType<typeof actions.createExpedition>
) => {
  return loop(
    state,
    Cmd.run(sideEffects.createExpeditionConfig, {
      args: [Cmd.getState, action.payload.baseConfig],
      successActionCreator: actions.createExpeditionSuccess,
    })
  )
}

export const createExpeditionSuccess = (
  state: State,
  action: ReturnType<typeof actions.createExpeditionSuccess>
) => {
  const newExpedition = action.payload

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [newExpedition.id]: newExpedition,
    },
    expeditionIds: [newExpedition.id, ...state.expeditionIds],
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const deleteExpedition = (
  state: State,
  action: ReturnType<typeof actions.deleteExpedition>
) => {
  const id = action.payload

  const { [id]: expedition, ...rest } = state.expeditions

  const newState = {
    expeditions: { ...rest },
    expeditionIds: state.expeditionIds.filter(
      expeditionId => expeditionId !== id
    ),
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const rollBattle = (
  state: State,
  action: ReturnType<typeof actions.rollBattle>
) => {
  const { battle, upgradedBasicNemesisCardIds } = action.payload

  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )
  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: battle,
  })

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        upgradedBasicNemesisCards: upgradedBasicNemesisCardIds,
        battles: updatedBattles,
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const startBattle = (
  state: State,
  action: ReturnType<typeof actions.startBattle>
) => {
  const { battle } = action.payload
  return updateBattle(state, battle, {
    tries: battle.tries + 1,
    status: 'started',
  })
}

export const winBattle = (
  state: State,
  action: ReturnType<typeof actions.winBattle>
) => {
  const { battle } = action.payload
  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )

  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: {
      ...battle,
      status: 'won',
    },
  })

  const battleScore = helpers.calcBattleScore(battle.tries)

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        score: oldExpedition.score + battleScore,
        battles: updatedBattles,
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const loseBattle = (
  state: State,
  action: ReturnType<typeof actions.loseBattle>
) => {
  const { battle } = action.payload
  return updateBattle(state, battle, { status: 'lost' })
}

export const rollLoss = (
  state: State,
  action: ReturnType<typeof actions.rollLoss>
) => {
  const battle = action.payload

  return updateBattle(state, battle)
}

export const acceptLoss = (
  state: State,
  action: ReturnType<typeof actions.acceptLoss>
) => {
  const { battle, banished, newSupplyIds } = action.payload

  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )

  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: {
      ...battle,
      status: 'before_battle',
      rewards: undefined,
    },
  })

  const newTreasureIds = battle.rewards ? battle.rewards.treasure : []
  const newMageIds =
    battle.rewards && battle.rewards.mage ? [battle.rewards.mage] : []

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        battles: updatedBattles,
        barracks: {
          ...oldExpedition.barracks,
          treasureIds: [
            ...oldExpedition.barracks.treasureIds,
            ...newTreasureIds,
          ],
          supplyIds: newSupplyIds,
          mageIds: [...oldExpedition.barracks.mageIds, ...newMageIds],
        },
        banished: [...oldExpedition.banished, ...banished],
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const finishBattle = (
  state: State,
  action: ReturnType<typeof actions.finishBattle>
) => {
  const { battle, newSupplyIds, banished } = action.payload
  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )

  const indexOfNextBattle = battleIndex + 1
  const hasNext = indexOfNextBattle < oldBattleList.length

  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: {
      ...battle,
      status: 'finished',
    },
    ...(hasNext && {
      [indexOfNextBattle]: {
        ...oldBattleList[indexOfNextBattle],
        status: 'unlocked',
      },
    }),
  })

  const newTreasureIds = battle.rewards ? battle.rewards.treasure : []

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        battles: updatedBattles,
        barracks: {
          ...oldExpedition.barracks,
          treasureIds: [
            ...oldExpedition.barracks.treasureIds,
            ...newTreasureIds,
          ],
          supplyIds: newSupplyIds,
        },
        banished: [...oldExpedition.banished, ...banished],
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const finishExpedition = (
  state: State,
  action: ReturnType<typeof actions.finishExpedition>
) => {
  const { battle } = action.payload
  const oldExpedition = state.expeditions[battle.expeditionId]
  const oldBattleList = oldExpedition.battles

  const battleIndex = oldBattleList.findIndex(
    oldBattle => oldBattle.id === battle.id
  )

  const updatedBattles = Object.assign([...oldBattleList], {
    [battleIndex]: {
      ...battle,
      status: 'finished',
    },
  })

  const battleScore = helpers.calcBattleScore(battle.tries)

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        score: oldExpedition.score + battleScore,
        battles: updatedBattles,
        finished: true,
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

export const fetchFromDB = (state: State, _: Action) => {
  return loop(
    state,
    Cmd.run(getFromDb, {
      args: [EXPEDITIONS_DB_KEY],
      successActionCreator: actions.fetchFromDBSuccessful,
      failActionCreator: actions.fetchFromDBFailed,
    })
  )
}

export const migrateAfterFetch = (
  action: ReturnType<typeof actions.fetchFromDBSuccessful>
) => {
  const expeditions = action.payload.expeditionIds.map(
    id => action.payload.expeditions[id]
  )
  const expeditionsToMigrate = expeditions.filter(expedition => {
    return !expedition.settingsSnapshot || !expedition.seed
  })

  if (expeditionsToMigrate.length === 0) {
    return action.payload
  }

  return loop(
    action.payload,
    Cmd.run(sideEffects.migrateToSettingsSnapshot, {
      args: [
        Cmd.getState,
        {
          state: action.payload,
          expeditionsToMigrate,
        },
      ],
      successActionCreator: actions.migrateToSettingsSnapshotSuccessful,
    })
  )
}

export const migrateAfterFetchSuccess = (
  state: State,
  action: ReturnType<typeof actions.migrateToSettingsSnapshotSuccessful>
) => {
  const newState = action.payload || state
  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
