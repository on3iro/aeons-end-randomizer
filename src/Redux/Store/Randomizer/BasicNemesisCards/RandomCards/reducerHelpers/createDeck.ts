import { State } from '../types'

import { actions } from '../actions'

import { getRandomBasicNemesisCardList } from './helpers'

export const createDeck = (
  state: State,
  action: ReturnType<typeof actions.createBasicNemesisCardDeck>
) => {
  const { availableCards, playerCount, seed } = action.payload

  const newState = getRandomBasicNemesisCardList(
    availableCards,
    playerCount,
    seed
  )

  return {
    ...state,
    Tier1: newState.Tier1,
    Tier2: newState.Tier2,
    Tier3: newState.Tier3,
  }
}
