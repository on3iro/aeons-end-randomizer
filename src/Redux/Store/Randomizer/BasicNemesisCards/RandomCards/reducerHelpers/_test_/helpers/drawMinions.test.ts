import * as types from 'types'

import { drawMinions } from 'Redux/Store/Randomizer/BasicNemesisCards/RandomCards/reducerHelpers/helpers'

const cards = [
  { id: 'card1', type: 'Minion' },
  { id: 'card2', type: 'Power' },
  { id: 'card3', type: 'Attack' },
  { id: 'card4', type: 'Power' },
  { id: 'card5', type: 'Minion' },
  { id: 'card6', type: 'Minion' },
] as types.BasicNemesisCard[]

describe('drawMinions()', () => {
  it('should return 1 minion', () => {
    const result = drawMinions(cards, 1, { seed: 'test' })

    expect(result).toMatchSnapshot()
  })

  it('should return 3 minions', () => {
    const result = drawMinions(cards, 3, { seed: 'test' })

    expect(result).toMatchSnapshot()
  })
})
