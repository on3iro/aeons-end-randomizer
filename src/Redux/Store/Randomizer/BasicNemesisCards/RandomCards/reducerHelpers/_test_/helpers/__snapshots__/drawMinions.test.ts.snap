// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawMinions() should return 1 minion 1`] = `
Array [
  Object {
    "id": "card6",
    "type": "Minion",
  },
]
`;

exports[`drawMinions() should return 3 minions 1`] = `
Array [
  Object {
    "id": "card6",
    "type": "Minion",
  },
  Object {
    "id": "card1",
    "type": "Minion",
  },
  Object {
    "id": "card5",
    "type": "Minion",
  },
]
`;
